### ISSUE TITLE
to_template_unique_name is not working as expected

### ISSUE THREAD

**Issue by thirtytwobits** [UAVCAN/nunavut#88]
The filter to_template_unique_name is malfunctioning here. Below is an example of misgenerated code where unique identifiers are not unique:

```
    @staticmethod
    def _deserialize_aligned_(_des_: Heartbeat_1_0._DeserializerTypeVar_) -> Heartbeat_1_0:
        assert _des_.consumed_bit_length % 8 == 0, 'Deserializer is not byte-aligned'
        _bit_length_base_ = _des_.consumed_bit_length
        _des_.require_remaining_bit_length(56)
        # BEGIN COMPOSITE DESERIALIZATION: uavcan.node.Heartbeat.1.0
        # BEGIN STRUCTURE FIELD DESERIALIZATION: saturated uint32 uptime
        # The temporary _f0_ holds the value of the field "uptime"
        _des_.require_remaining_bit_length(32)
        assert _des_.consumed_bit_length % 8 == 0, 'saturated uint32'
        _f0_ = _des_.fetch_aligned_u32()
        # END STRUCTURE FIELD DESERIALIZATION: saturated uint32 uptime
        # BEGIN STRUCTURE FIELD DESERIALIZATION: truncated uint2 health
        # The temporary _f0_ holds the value of the field "health"
        _des_.require_remaining_bit_length(2)
        assert _des_.consumed_bit_length % 8 == 0, 'truncated uint2'
        _f0_ = _des_.fetch_aligned_unsigned(2)
        # END STRUCTURE FIELD DESERIALIZATION: truncated uint2 health
        # BEGIN STRUCTURE FIELD DESERIALIZATION: truncated uint3 mode
        # The temporary _f0_ holds the value of the field "mode"
        _des_.require_remaining_bit_length(3)
        _f0_ = _des_.fetch_unaligned_unsigned(3)
        # END STRUCTURE FIELD DESERIALIZATION: truncated uint3 mode
        # BEGIN STRUCTURE FIELD DESERIALIZATION: truncated uint19 vendor_specific_status_code
        # The temporary _f0_ holds the value of the field "vendor_specific_status_code"
        _des_.require_remaining_bit_length(19)
        _f0_ = _des_.fetch_unaligned_unsigned(19)
        # END STRUCTURE FIELD DESERIALIZATION: truncated uint19 vendor_specific_status_code
        self = Heartbeat_1_0(uptime=_f0_,
                             health=_f0_,
                             mode=_f0_,
                             vendor_specific_status_code=_f0_)
```

https://github.com/UAVCAN/pyuavcan/blob/6f1544763d7c1b1baa688d1cf63d35c3818ef53e/pyuavcan/dsdl/_templates/deserialization.j2#L83

**Comment by pavel-kirienko** [comment]
We are dealing with a Heisenbug! Outside of PyUAVCAN it works properly.

```
⋊> ~/test cat nunavut/UnionType.j2
{%- include 'StructureType.j2' -%}

⋊> ~/test cat nunavut/ServiceType.j2
{%- include 'StructureType.j2' -%}

⋊> ~/test cat nunavut/StructureType.j2
{%- set field_ref = 'f'|to_template_unique_name %}
{{ field_ref }}
{%- set field_ref = 'f'|to_template_unique_name %}
{{ field_ref }}
{{ 'f'|to_template_unique_name }}

⋊> ~/test nnvg ~/uavcan/public_regulated_data_types/uavcan --templates nunavut --output-extension=py --target-language=py 

⋊> ~/test cat nunavut_out/uavcan/time/SynchronizedTimestamp_1_0.py

_f0_
_f1_
_f2_
```

**Comment by pavel-kirienko** [comment]
When I replace the above dummy templates with those from PyUAVCAN, the bug is back:

```
⋊> ~/test nnvg -vv ~/uavcan/public_regulated_data_types/uavcan --templates ~/uavcan/pyuavcan/pyuavcan/dsdl/_templates/ --output-extension=py --target-language=py 2> nunavut.log
⋊> ~/test cat nunavut_out/uavcan/diagnostic/Record_1_0.py | grep '_f._' 
        # The temporary _f0_ holds the value of the field "timestamp"
        _f0_ = uavcan.time.SynchronizedTimestamp_1_0._deserialize_aligned_(_des_)
        # The temporary _f0_ holds the value of the field "severity"
        _f0_ = uavcan.diagnostic.Severity_1_0._deserialize_aligned_(_des_)
        # The temporary _f0_ holds the value of the field "text"
        _f0_ = _des_.fetch_aligned_array_of_standard_bit_length_primitives(, _len0_)
        assert len(_f0_) <= 112, 'saturated uint8[<=112]'
        self = Record_1_0(timestamp=_f0_,
                          severity=_f0_,
                          text=_f0_)
```

To make the above test work, I had to drop the following additional primitives from my templates because they are not available in Nunavut, but this change does not affect the bug:

```python
    # Template primitives
    filters = {
        'pickle':            _pickle_object,
        'numpy_scalar_type': _numpy_scalar_type,
    }

    tests = {
        'PaddingField': lambda x: isinstance(x, pydsdl.PaddingField),
        'saturated':    _test_if_saturated,
    }
```

**Comment by pavel-kirienko** [comment]
Logs :evergreen_tree: 

`uvc -vv dsdl-gen-pkg ../uavcan/public_regulated_data_types/uavcan/ -O pyuavcan 2> pyuavcan.log` [pyuavcan.log](https://github.com/UAVCAN/nunavut/files/3778234/pyuavcan.log)

`nnvg -vv ~/uavcan/public_regulated_data_types/uavcan --templates ~/uavcan/pyuavcan/pyuavcan/dsdl/_templates/ --output-extension=py --target-language=py 2> nunavut-pyuavcan.log` [nunavut-pyuavcan.log](https://github.com/UAVCAN/nunavut/files/3778239/nunavut-pyuavcan.log)

**Comment by pavel-kirienko** [comment]
**BEHOLD THE MWE**

```
{%- macro suka() -%}
{{ 'f'|to_template_unique_name }}
{{ 'f'|to_template_unique_name }}
{{ 'f'|to_template_unique_name }}
{{ 'f'|to_template_unique_name }}
{%- endmacro -%}

{{ 'f'|to_template_unique_name }}
{{ 'f'|to_template_unique_name }}
{{ 'f'|to_template_unique_name }}
{{ 'f'|to_template_unique_name }}

{{ suka() }}

{{ suka() }}
```

Expected output:

```
_f0_
_f1_
_f2_
_f3_

_f4_
_f5_
_f6_
_f7_

_f8_
_f9_
_f10_
_f11_
```

Actual output:

```
_f4_
_f5_
_f6_
_f7_

_f0_
_f1_
_f2_
_f3_

_f0_
_f1_
_f2_
_f3_
```

The MWE is very fragile and even a minor modification may "fix" it.

In the output, you can see that the top-level statements have been executed *after the first macro invocation*, which obviously doesn't make any sense at all. This is either a glitch in the Matrix or there is some optimization nonsense going on that makes wrong assumptions about the logic. If the cache is disabled it works as it should:

```diff
------------------------ src/nunavut/jinja/__init__.py ------------------------
index fcd7792..4753985 100644
@@ -404,7 +404,8 @@ class Generator(nunavut.generators.AbstractGenerator):
                                 keep_trailing_newline=True,
                                 lstrip_blocks=lstrip_blocks,
                                 trim_blocks=trim_blocks,
-                                auto_reload=False)
+                                auto_reload=False,
+                                cache_size=0)

         self._add_language_support()
```

This tells us that either Jinja is trying to be too clever (making a strong assumption that macros are pure functions, in our case they are not) or we are using it incorrectly. I am perfectly fine with just disabling the cache, but then we should ask ourselves why are we even using Jinja in the first place.

**Comment by thirtytwobits** [comment]
>  I am perfectly fine with just disabling the cache

Yeah, me too. So, that fixes your bug then?

>  but then we should ask ourselves why are we even using Jinja in the first place

That's not the right question, in my view. What we should have asked ourselves is why we need a templating language. I'm still convinced that Nunavut should be rewritten in the future removing all templating and moving to a more compiler-like design but It shouldn't actually matter all that much by then since we are moving towards the interfaces being the nnvg CLI and the generated code itself.

**Comment by thirtytwobits** [comment]
Also, can you try just setting `bytecode_cache` to 0 and seeing if that works?

**Comment by thirtytwobits** [comment]
Hmm. From https://jinja.palletsprojects.com/en/2.10.x/templates/#import-visibility

> Import Context Behavior
> By default, included templates are passed the current context and imported templates are not. The reason for this is that imports, unlike includes, are cached; as imports are often used just as a module that holds macros.

**Comment by pavel-kirienko** [comment]
> Also, can you try just setting bytecode_cache to 0 and seeing if that works?

It doesn't because `bytecode_cache` should be either a bytecode cache instance or None.

> So, that fixes your bug then?

Yes.

**Comment by pavel-kirienko** [comment]
Fixed in #100.
